Clarify inference example docs

damccorm opened the pull request:
I walked through the inference examples - they're great, but I wanted to try to clean up some points of friction, confusion, or inconsistency that I found and streamline them a bit. Note, this is a little opinionated, definitely feel free to disagree if you think this makes things less clear in places.

Also removes the skip_header_lines arg when we're reading from txt files. I didn't get why we had that, but if there's a good reason we should doc it in the README.

Fixes #23056

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

damccorm commented:
R: @yeandy @rszper

rszper commented on sdks/python/apache_beam/examples/inference/README.md:
```diff
@@ -24,7 +24,7 @@ API. <!---TODO: Add link to full documentation on Beam website when it's publish
 ## Prerequisites
 
-You must have `apache-beam>=2.40.0` installed in order to run these pipelines,
+You must have `apache-beam>=2.40.0` or greater installed in order to run these pipelines,
```
greater -> later

rszper commented:
LGTM, one minor comment

bvolpato commented on sdks/python/apache_beam/examples/inference/README.md:
````diff
@@ -67,37 +67,35 @@ The pipeline reads the images, performs basic preprocessing, passes the images t
 One popular dataset is from [ImageNet](https://www.image-net.org/). Follow their instructions to download the images.
-2. Create a file named `IMAGE_FILE_NAMES` that contains the absolute paths of each of the images in `IMAGES_DIR` that you want to use to run image classification. The path to the file can be different types of URIs such as your local file system, an AWS S3 bucket, or a GCP Cloud Storage bucket. For example:
+2. Create a file named `IMAGE_FILE_NAMES.txt` that contains the absolute paths of each of the images in `IMAGES_DIR` that you want to use to run image classification. The path to the file can be different types of URIs such as your local file system, an AWS S3 bucket, or a GCP Cloud Storage bucket. For example:
 ```
 /absolute/path/to/image1.jpg
 /absolute/path/to/image2.jpg
 ```
````
nit: Having a `gs://bucket/path/to/image3.jpg` would emphasize the ability to use images directly from Cloud Storage, as well as show the proper format to do so.

damccorm replied:
I actually think that leading with local is a better experience here. Most people are going to want the simplest example path.

bvolpato commented on sdks/python/apache_beam/examples/inference/README.md:
````diff
@@ -24,7 +24,7 @@ API. <!---TODO: Add link to full documentation on Beam website when it's publish
+You must have `apache-beam>=2.40.0` or greater installed in order to run these pipelines,
 because the `apache_beam.examples.inference` module was added in that release.
 ```
 pip install apache-beam==2.40.0
````
Should we have `apache-beam>=2.40.0` to be consistent with the above?

damccorm replied:
Good call, I updated to do this

bvolpato commented on sdks/python/apache_beam/examples/inference/README.md:
```diff
@@ -231,16 +229,15 @@ The pipeline reads rows of pixels corresponding to a digit, performs basic prepr
 To use this transform, you need a dataset and model for language modeling.
 
-1. Create a file named `INPUT` that contains labels and pixels to feed into the model. Each row should have comma-separated elements. The first element is the label. All other elements are pixel values. The content of the file should be similar to the following example:
+1. Create a file named `INPUT.csv` that contains labels and pixels to feed into the model. Each row should have comma-separated elements. The first element is the label. All other elements are pixel values. The content of the file should be similar to the following example:
```
+1, adding the extension can go a long way to show expected format and improve tools usage when following the example.

AnandInguva commented:
Run Python 3.9 PostCommit

AnandInguva commented on sdks/python/apache_beam/examples/inference/sklearn_mnist_classification.py:
```diff
@@ -104,8 +104,7 @@ def run(
   label_pixel_tuple = (
       pipeline
-      | "ReadFromInput" >> beam.io.ReadFromText(
-          known_args.input, skip_header_lines=1)
```
skip_header_lines is used, just in case if the csv file has column names.

damccorm replied:
I think its a pretty surprising experience, especially since we don't have headers in our example csv. I called out that the created csv shouldn't have headers though.

AnandInguva commented:
Run Python 3.7 PostCommit

bvolpato commented on sdks/python/apache_beam/examples/inference/README.md:
````diff
@@ -67,37 +67,35 @@ The pipeline reads the images, performs basic preprocessing, passes the images t
 To run the image classification pipeline locally, use the following command:
 ```sh
 python -m apache_beam.examples.inference.pytorch_image_classification \
   --input IMAGE_FILE_NAMES \
````
```suggestion
  --input IMAGE_FILE_NAMES.txt \
```

bvolpato commented on sdks/python/apache_beam/examples/inference/README.md:
```diff
@@ -67,37 +67,35 @@ The pipeline reads the images, performs basic preprocessing, passes the images t
 python -m apache_beam.examples.inference.pytorch_image_classification \
   --input IMAGE_FILE_NAMES \
-  --images_dir IMAGES_DIR \
+  --images_dir IMAGES_DIR \ # Only needed if your IMAGE_FILE_NAMES file contains relative paths (they will be relative from IMAGES_DIR).
```
Pull the comment elsewhere or add `\` after it. Can't run the shell as-is now

damccorm replied:
Good call - I moved all the comments out of the script blocks (here and elsewhere).

bvolpato commented on sdks/python/apache_beam/examples/inference/README.md:
````diff
@@ -192,14 +190,14 @@ He looked up and saw the sun and stars .
 ```sh
 python -m apache_beam.examples.inference.pytorch_language_modeling \
-  --input sentences.txt \
+  --input SENTENCES.txt \ # optional
````
Move the comment or add `\` to the end?

bvolpato commented on sdks/python/apache_beam/examples/inference/README.md:
```diff
@@ -218,8 +216,8 @@ He looked up and saw the sun and stars .;moon
+the second item is the word that the model predicts for the mask.
 
 ---
 ## MNITST digit classification
```
```suggestion
## MNIST digit classification
```

bvolpato reviewed (changes requested):
Added a few comments, can you take a look and see if they make sense?

My friction log, in case it's helpful:

### Transformers Dependencies

`pip install transformers` required Rust installed:
	error: can't find Rust compiler

Probably worth to mention as a prerequisite: https://www.rust-lang.org/tools/install

----

### Image files being ignored?

Apparently, the first line of my IMAGE_FILE_NAMES.txt is being ignored all the time.

----

### Pipeline is too sensible to basic user errors

I had an extra blank line on my SENTENCES.txt file, and there were some weird failures. Likely could be ignored:
```
  File "/Users/bvolpato/.pyenv/versions/3.9.13/lib/python3.9/site-packages/apache_beam/examples/inference/pytorch_language_modeling.py", line 91, in add_mask_to_last_word
    return text, ' '.join(text_list[:-2] + ['[MASK]', text_list[-1]])
IndexError: list index out of range [while running 'AddMask’]
```

----

### Maintainability

IMHO `pip install torchvision/transformers` could have a fixed version as well, so users don’t run into problems if those dependencies make incompatible changes.

damccorm commented:
> pip install transformers required Rust installed:
> error: can't find Rust compiler
>
> Probably worth to mention as a prerequisite: https://www.rust-lang.org/tools/install

I didn't run into this and I don't have rust installed so this is surprising - maybe its dependent on the version that gets installed? Its surprising, but I'm inclined to provide the minimal set that needs to be installed so that we don't need to keep that up to date (though if you convince me to pin to a specific version that changes 😄 )

> Apparently, the first line of my IMAGE_FILE_NAMES.txt is being ignored all the time.

Are you using my branch? I made a change to fix this

> I had an extra blank line on my SENTENCES.txt file, and there were some weird failures. Likely could be ignored:

I agree this should be fixed, I'll get it tomorrow morning

> IMHO pip install torchvision/transformers could have fixed versions as well, so users don’t run into problems if those dependencies make incompatible changes.

I think I disagree - Beam should work with a range of versions of these deps, and if it doesn't (beyond what pip enforces) its probably concerning (especially because we're not doing things w/ much complexity here). I like the idea of making that clear to users as well. I'm open to being convinced on this point.

bvolpato commented:
> I didn't run into this and I don't have rust installed so this is surprising - maybe its dependent on the version that gets installed? Its surprising, but I'm inclined to provide the minimal set that needs to be installed so that we don't need to keep that up to date (though if you convince me to pin to a specific version that changes 😄 )

Do you get something on `which rustc`? Just doing `pip install transformers` was failing for me with the above error.
Apparently it comes from https://github.com/huggingface/tokenizers, which is built on Rust.


> Are you using my branch? I made a change to fix this

My bad. I followed the README from your branch but was using the Beam module straight from the `pip install`.

> I think I disagree - Beam should work with a range of versions of these deps, and if it doesn't (beyond what pip enforces) its probably concerning (especially because we're not doing things w/ much complexity here). I like the idea of making that clear to users as well. I'm open to being convinced on this point.

I don't have a strong opinion here. I mentioned so we can be consistent with the installation of torch, and it feels that not specifying a version may put the burden on us if those libraries break something. 
I'm not familiar with all the tests/stack here yet, but could we have some automation to go through steps on this tutorial from time to time? So at least we get to know before anyone reports.

AnandInguva commented:
I updated the csv files and removed the column names from csv since the examples won't use them. The IT tests passes now

AnandInguva commented:
Run Python 3.7 PostCommit

AnandInguva commented:
Run Python 3.9 PostCommit

AnandInguva commented:
> > I didn't run into this and I don't have rust installed so this is surprising - maybe its dependent on the version that gets installed? Its surprising, but I'm inclined to provide the minimal set that needs to be installed so that we don't need to keep that up to date (though if you convince me to pin to a specific version that changes 😄 )
> 
> Do you get something on `which rustc`? Just doing `pip install transformers` was failing for me with the above error (Mac M1, unsure if matters). Apparently it comes from https://github.com/huggingface/tokenizers, which is built on Rust.
> 
> > Are you using my branch? I made a change to fix this
> 
> My bad. I followed the README from your branch but was using the Beam module straight from the `pip install`.
> 
> > I think I disagree - Beam should work with a range of versions of these deps, and if it doesn't (beyond what pip enforces) its probably concerning (especially because we're not doing things w/ much complexity here). I like the idea of making that clear to users as well. I'm open to being convinced on this point.
> 
> I don't have a strong opinion here. I mentioned so we can be consistent with the installation of torch, and it feels that not specifying a version may put the burden on us if those libraries break something. I'm not familiar with all the tests/stack here yet, but could we have some automation to go through steps on examples/tutorials from time to time? So at least we get to know before anyone reports, or even worse, avoid that someone can't get it working and doesn't report at all.
> 
> To add on the last sentence, can we somehow encourage people to file an issue/report any problems or frictions that they have while going over the examples?

@BjornPrime also encountered something related to Rust. May be he can provide some context if its related to Beam or Hugging face.

damccorm commented:
@bvolpato 

> Do you get something on which rustc?

Nope - weird that you're seeing this and I'm not though. Which version of transformers is pip giving you (`pip show transformers`)? I'm on 4.21.2.

Regardless, I'm probably not inclined to do too much here since this is out of our control (and other dependencies could add additional requirements).

> I'm not familiar with all the tests/stack here yet, but could we have some automation to go through steps on examples/tutorials from time to time?

I think the postcommits will at least make sure they compile correctly, doing a good dependency test here is pretty tough though because it is a little environmentally dependent (for example, I'm getting the error, you're not :) ). We'll also probably get different results depending on the platform.

---------

Also, updated the examples to filter out empty lines

damccorm commented:
> I updated the csv files and removed the column names from csv since the examples won't use them. The IT tests passes now

@AnandInguva thanks - where do those files live for future reference?

damccorm commented:
Run Python 3.7 PostCommit

damccorm commented:
Run Python 3.9 PostCommit

yeandy commented on sdks/python/apache_beam/examples/inference/pytorch_image_segmentation.py:
```diff
@@ -225,8 +231,8 @@ def run(
-      | 'ReadImageNames' >> beam.io.ReadFromText(
-          known_args.input, skip_header_lines=1)
+      | 'ReadImageNames' >> beam.io.ReadFromText(known_args.input)
+      | 'RemoveEmptyLines' >> beam.ParDo(filter_empty_lines)
```
```suggestion
      | 'FilterEmptyLines' >> beam.ParDo(filter_empty_lines)
```

yeandy commented on sdks/python/apache_beam/examples/inference/pytorch_image_classification.py:
```diff
@@ -136,8 +142,8 @@ def run(
-      | 'ReadImageNames' >> beam.io.ReadFromText(
-          known_args.input, skip_header_lines=1)
+      | 'ReadImageNames' >> beam.io.ReadFromText(known_args.input)
+      | 'RemoveEmptyLines' >> beam.ParDo(filter_empty_lines)
```
```suggestion
      | 'FilterEmptyLines' >> beam.ParDo(filter_empty_lines)
```

yeandy commented on sdks/python/apache_beam/examples/inference/README.md:
````diff
@@ -67,22 +67,20 @@ The pipeline reads the images, performs basic preprocessing, passes the images t
 /absolute/path/to/image2.jpg
 ```
-3. Download the [mobilenet_v2](https://pytorch.org/vision/stable/_modules/torchvision/models/mobilenetv2.html) model from Pytorch's repository of pretrained models. This model requires the torchvision library. To download this model, run the following commands:
+3. Download the [mobilenet_v2](https://pytorch.org/vision/stable/_modules/torchvision/models/mobilenetv2.html) model from Pytorch's repository of pretrained models. This model requires the torchvision library. To download this model, run the following commands from a python shell:
````
```suggestion
3. Download the [mobilenet_v2](https://pytorch.org/vision/stable/_modules/torchvision/models/mobilenetv2.html) model from Pytorch's repository of pretrained models. This model requires the torchvision library. To download this model, run the following commands from a Python shell:
```

yeandy commented on sdks/python/apache_beam/examples/inference/README.md:
````diff
@@ -94,10 +92,12 @@ python -m apache_beam.examples.inference.pytorch_image_classification \
   --model_state_dict_path MODEL_STATE_DICT
 ```
-For example:
+`images_dir` is only needed if your `IMAGE_FILE_NAMES` file contains relative paths (they will be relative from `IMAGES_DIR`).
````
```suggestion
`images_dir` is only needed if your `IMAGE_FILE_NAMES.txt` file contains relative paths (they will be relative from `IMAGES_DIR`).
```

yeandy commented on sdks/python/apache_beam/examples/inference/README.md:
````diff
@@ -118,22 +121,21 @@ The pipeline reads images, performs basic preprocessing, passes the images to th
 /absolute/path/to/image2.jpg
 ```
-3. Download the [maskrcnn_resnet50_fpn](https://pytorch.org/vision/0.12/models.html#id70) model from Pytorch's repository of pretrained models. This model requires the torchvision library. To download this model, run the following commands:
+3. Download the [maskrcnn_resnet50_fpn](https://pytorch.org/vision/0.12/models.html#id70) model from Pytorch's repository of pretrained models. This model requires the torchvision library. To download this model, run the following commands from a python shell:
````
```suggestion
3. Download the [maskrcnn_resnet50_fpn](https://pytorch.org/vision/0.12/models.html#id70) model from Pytorch's repository of pretrained models. This model requires the torchvision library. To download this model, run the following commands from a Python shell:
```

yeandy commented on sdks/python/apache_beam/examples/inference/README.md:
````diff
@@ -145,10 +147,12 @@ python -m apache_beam.examples.inference.pytorch_image_segmentation \
   --model_state_dict_path MODEL_STATE_DICT
 ```
-For example:
+`images_dir` is only needed if your `IMAGE_FILE_NAMES` file contains relative paths (they will be relative from `IMAGES_DIR`).
````
```suggestion
`images_dir` is only needed if your `IMAGE_FILE_NAMES.txt` file contains relative paths (they will be relative from `IMAGES_DIR`).
```

yeandy commented on sdks/python/apache_beam/examples/inference/README.md:
```diff
@@ -171,16 +175,14 @@ The pipeline reads sentences, performs basic preprocessing to convert the last w
 To use this transform, you need a dataset and model for language modeling.
 
-1. Download the [BertForMaskedLM](https://huggingface.co/docs/transformers/model_doc/bert#transformers.BertForMaskedLM) model from Hugging Face's repository of pretrained models. You must already have `transformers` installed.
+1. Download the [BertForMaskedLM](https://huggingface.co/docs/transformers/model_doc/bert#transformers.BertForMaskedLM) model from Hugging Face's repository of pretrained models. You must already have `transformers` installed, then from a python shell run:
```
```suggestion
1. Download the [BertForMaskedLM](https://huggingface.co/docs/transformers/model_doc/bert#transformers.BertForMaskedLM) model from Hugging Face's repository of pretrained models. You must already have `transformers` installed, then from a Python shell run:
```

yeandy commented on sdks/python/apache_beam/examples/inference/README.md:
````diff
@@ -67,22 +67,20 @@ The pipeline reads the images, performs basic preprocessing, passes the images t
-torch.save(model.state_dict(), 'mobilenet_v2.pth')
+torch.save(model.state_dict(), 'mobilenet_v2.pth') # You can replace mobilenet_v2.pth with your preferred file name for your model state dictionary.
 ```
-4. Create a file named `MODEL_STATE_DICT` that contains the saved parameters of the `mobilenet_v2` model.
-5. Note the path to the `OUTPUT` file. This file is used by the pipeline to write the predictions.
````
Why did you remove comments about `MODEL_STATE_DICT` and `OUTPUT`?

damccorm replied:
`MODEL_STATE_DICT` has been created as part of a previous step, the `OUTPUT` line just confused me since it doesn't exist yet. The example command below is clearer IMO.

yeandy replied:
👍

yeandy commented on sdks/python/apache_beam/examples/inference/README.md:
````diff
@@ -118,22 +121,21 @@ The pipeline reads images, performs basic preprocessing, passes the images to th
-torch.save(model.state_dict(), 'maskrcnn_resnet50_fpn.pth')
+torch.save(model.state_dict(), 'maskrcnn_resnet50_fpn.pth') # You can replace maskrcnn_resnet50_fpn.pth with your preferred file name for your model state dictionary.
 ```
-4. Create a path to a file named `MODEL_STATE_DICT` that contains the saved parameters of the `maskrcnn_resnet50_fpn` model.
````
Why did you remove comments about `MODEL_STATE_DICT`?

yeandy commented on sdks/python/apache_beam/examples/inference/README.md:
````diff
@@ -171,16 +175,14 @@ The pipeline reads sentences, performs basic preprocessing to convert the last w
-torch.save(model.state_dict(), 'BertForMaskedLM.pth')
+torch.save(model.state_dict(), 'BertForMaskedLM.pth') # You can replace BertForMaskedLM.pth with your preferred file name for your model state dictionary.
 ```
-2. Create a file named `MODEL_STATE_DICT` that contains the saved parameters of the `BertForMaskedLM` model.
````
Why did you remove comments about `MODEL_STATE_DICT` and `OUTPUT`?

damccorm replied:
Same as above

AnandInguva commented:
> > I updated the csv files and removed the column names from csv since the examples won't use them. The IT tests passes now
> 
> @AnandInguva thanks - where do those files live for future reference?

they live at gs://apache-beam-ml/testing/inputs.

AnandInguva commented:
This PR also fixes https://github.com/apache/beam/issues/23056
